pkg/erofs/erofs.go

```diff
@@ -0,0 +1,729 @@
+func (i *Image) RootNid() uint64 {
+	return uint64(i.sb.RootNid)
+}
+
+// readSuperBlock reads the super block of this image.
+func (i *Image) readSuperBlock() error {
+	if err := i.UnmarshalAt(&i.sb, SuperBlockOffset); err != nil {
+		return fmt.Errorf("image size is too small")
+	}
+
+	if i.sb.Magic != SuperBlockMagicV1 {
+		return fmt.Errorf("unknown magic: 0x%x", i.sb.Magic)
+	}
+
+	if err := i.verifyChecksum(); err != nil {
+		return err
+	}
+
+	if featureIncompat := i.sb.FeatureIncompat & ^uint32(FeatureIncompatSupported); featureIncompat != 0 {
+		return fmt.Errorf("unsupported incompatible features detected: 0x%x", featureIncompat)
+	}
+
+	if i.BlockSize()%hostarch.PageSize != 0 {
+		return fmt.Errorf("unsupported block size: 0x%x", i.BlockSize())
+	}
+
+	return nil
+}
+
+// verifyChecksum verifies the checksum of the super block.
+func (i *Image) verifyChecksum() error {
+	if i.sb.FeatureCompat&FeatureCompatSuperBlockChecksum == 0 {
+		return nil
+	}
+
+	sb := i.sb
+	sb.Checksum = 0
+	table := crc32.MakeTable(crc32.Castagnoli)
+	checksum := crc32.Checksum(marshal.Marshal(&sb), table)
+
+	off := SuperBlockOffset + uint64(i.sb.SizeBytes())
+	if bytes, err := i.BytesAt(off, uint64(i.BlockSize())-off); err != nil {
+		return fmt.Errorf("image size is too small")
+	} else {
+		checksum = ^crc32.Update(checksum, table, bytes)
+	}
+	if checksum != i.sb.Checksum {
+		return fmt.Errorf("invalid checksum: 0x%x, expected: 0x%x", checksum, i.sb.Checksum)
+	}
+
+	return nil
+}
+
+// FD returns the host FD of underlying image file.
+func (i *Image) FD() int {
+	return int(i.src.Fd())
+}
+
+// BytesAt returns the bytes at [off, off+n) of the image.
+func (i *Image) BytesAt(off, n uint64) ([]byte, error) {
+	size := uint64(len(i.bytes))
+	end := off + n
+	if off >= size || off > end || end > size {
+		log.Warningf("Invalid range (off: 0x%x, n: 0x%x) for image (size: 0x%x)", off, n, size)
+		return nil, linuxerr.EFAULT
+	}
+	return i.bytes[off:end], nil
+}
+
+// UnmarshalAt deserializes data from the bytes at [off, off+n) of the image.
+func (i *Image) UnmarshalAt(data marshal.Marshallable, off uint64) error {
+	bytes, err := i.BytesAt(off, uint64(data.SizeBytes()))
+	if err != nil {
+		log.Warningf("Failed to deserialize %T from 0x%x.", data, off)
+		return err
+	}
+	data.UnmarshalUnsafe(bytes)
+	return nil
+}
+
+// Inode returns the inode identified by nid.
+//
+// TODO: Ideally, we should avoid escaping objects to heap when constructing
+// objects from the image.
+func (i *Image) Inode(nid uint64) (Inode, error) {
+	inode := Inode{
+		image: i,
+		nid:   nid,
+	}
+
+	off := i.sb.NidToOffset(nid)
+	if err := i.UnmarshalAt(&inode.format, off); err != nil {
+		return Inode{}, err
+	}
+
+	var (
+		rawBlockAddr uint32
+		inodeSize    int
+	)
+
+	switch layout := inode.Layout(); layout {
+	case InodeLayoutCompact:
+		var ino InodeCompact
+		if err := i.UnmarshalAt(&ino, off); err != nil {
+			return Inode{}, err
+		}
```

Comment:
Thanks for the details and pointers! They are very helpful!

Hmm.. I prefer to have a simple and clean implementation as the initial support first, and add more sophisticated optimizations (which might be dirty or not very straightforward) on top.. Can I do this optimization in a separate PR or patch? Thanks!

Apart from the escape, I also plan to avoid the unnecessary memory copies. In most cases, we don't really need to unmarshal the objects and ideally we should generate a pointer that points to the object in the memory backed by the image file directly to replace the unmarshal operations.

Reply:
> Apart from that, there are no other functions that take interface parameters in EROFS right?

Yeah, there are no other functions defined by EROFS that take interface parameters.

Reply:
> Can I do this optimization in a separate PR or patch?

Sounds good.

> we don't really need to unmarshal the objects

So like just do an unsafe type cast (like C++ reinterpret_cast<>()) to the required struct type? Hmm I guess that could work.
Note that UnmarshalUnsafe() operation is really cheap. It uses `unsafe` package and just does a `memmove()` into the struct's memory.

Reply:
> So like just do an unsafe type cast (like C++ reinterpret_cast<>()) to the required struct type? Hmm I guess that could work.

Yeah.

> Note that UnmarshalUnsafe() operation is really cheap. It uses unsafe package and just does a memmove() into the struct's memory.

Yeah, I agree. In above case, there is no huge memory copies. And copying few small objects to the temp buffer in stack will not increase the memory footprint obviously. The overhead should be fine. I will see if I can create a clean implementation based on the "cast" approach which is supposed to have zero overhead first. And if not, I will take the optimized unmarshal approach.